docs/architecture/adr-034-account-rekeying.md

```diff
@@ -0,0 +1,75 @@
+
+## Changelog
+
+- 30-09-2020: Initial Draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+Account rekeying is a process hat allows an account to replace its authentication pubkey with a new one.
+
+## Context
+
+Currently, in the Cosmos SDK, the address of an auth `BaseAccount` is based on the hash of the public key.  Once an account is created, the public key for the account is set in stone, and cannot be changed.  This can be a problem for users, as key rotation is a useful security practice, but is not possible currently.  Furthermore, as multisigs are a type of pubkey, once a multisig for an account is set, it can not be updated.  This is problematic, as multisigs are often used by organizations or companies, who may need to change their set of multisig signers for internal reasons.
+
+Transferring all the assets of an account to a new account with the updated pubkey is not sufficient, because some "engagements" of an account are not easily transferable.  For example, in staking, to transfer bonded Atoms, an account would have to unbond all delegations and wait the three week unbonding period.  Even more significantly, for validator operators, ownership over a validator is not transferrable at all, meaning that the operator key for a validator can never be updated, leading to poor operational security for validators. 
+
+## Decision
+
+We propose the addition of a new feature to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.  This feature can be enabled using an `EnableChangePubKey` param.
+
+This is possible because the Cosmos SDK `BaseAccount` stores the public key for an account in state, instead of making the assumption that the public key is included in the transaction (whether explicitly or implicitly through the signature) as in other blockchains such as Bitcoin and Ethereum.  Because the public key is stored on chain, it is okay for the public key to not hash to the address of an account, as the address is not pertinent to the signature checking process.
+
+To build this system, we design a new Msg type as follows:
+
+```protobuf
+service Msg {
+    rpc ChangePubKey(MsgChangePubKey) returns (MsgChangePubKeyResponse);
+}
+
+message MsgChangePubKey {
+  string address = 1;
+  google.protobuf.Any pub_key = 2;
+}
+
+message MsgChangePubKeyResponse {}
+```
+
+The MsgChangePubKey transaction needs to be signed by the existing pubkey in state.
+
+Once, approved, the handler for this message type, which takes in the AccountKeeper, will update the in-state pubkey for the account and replace it with the pubkey from the Msg.
+
+Because an account can no longer be pruned from state once its pubkey has changed, we can charge an additional gas fee for this operation to compensate for this this externality (this bound gas amount is configured as parameter `PubKeyChangeCost`). The bonus gas is charged inside the handler, using the `ConsumeGas` function.
```

Comment:
`Because an account can no longer be pruned from state once its pubkey has changed...`. Can you please elaborate on this.

Reply:
Check this comment, and the link: https://github.com/cosmos/cosmos-sdk/pull/7491#discussion_r519160750

It's about remembering which address is used with which PK. I believe this is the scenario:
1. you have `account1 = {pk1}`
2. your `(pk1, sk1)` was compromised
3. you rekey: `account1 = {pk2}`
4. we want to assure that `pk1`  won't be used in the future. (otherwise it could do actions for `account1`, even if it was deleted).

Maybe there are cases, which will be good to elaborate.

Reply:
Yes, I'm aware of this situation -- I just meant the ADR should describe it because not everyone has this context 👍